### PR TITLE
refactor: rewrite Typesystem without context propagation

### PR DESCRIPTION
**Pull request by jon-whit**
## Description
Refactors the way the Typesystem is resolved and passed around. Instead of passing through context we explicitly propagate it via a TypesystemResolverFunc that resolves it in a memoized way.

## References
https://github.com/openfga/openfga/issues/616

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/typesystem/resolver_test.go
Comment (jon-whit) [inline pkg/typesystem/resolver_test.go:15]: todo(@jon-whit) - You should also add one more test for the case where the authorization model identifier is `""`.

File: pkg/server/commands/connected_objects.go
Comment (jon-whit) [inline pkg/server/commands/connected_objects.go:105]: Function signature is starting to get pretty bloated. Will consider some ways to make this a bit more elegant, but this works at least for the objective at hand.

File: pkg/server/queries.go
Comment (jon-whit) [inline pkg/server/queries.go:1]: I added the `queries.go` file to separate out the Relationship Query RPC handlers.

https://openfga.dev/api/service#/Relationship%20Queries

**Comment by jon-whit** [comment]
Looks like something broke in the tests. I'll investigate next week 👍

File: pkg/server/commands/connected_objects.go
Comment (miparnisari) [reply pkg/server/commands/connected_objects.go:105]: Why can't `typesys` be kept as part of `ConnectedObjectsRequest` and `reverseExpandRequest`?

File: pkg/server/commands/list_objects.go
Comment (miparnisari) [inline pkg/server/commands/list_objects.go:53]: Similar comment as above. `typesys` is a concrete instance of a model ID. It can be part of `listObjectsRequest`

**Review by miparnisari — CHANGES_REQUESTED** [review]

File: pkg/server/queries.go
Comment (miparnisari) [reply pkg/server/queries.go:1]: Please don't move code around in this PR

File: pkg/server/commands/connected_objects.go
Comment (jon-whit) [reply pkg/server/commands/connected_objects.go:105]: Our pattern is clear - `<something>Request` structs are for user provided inputs. The typesys is an structure constructed internally during the request lifecycle. The suggestion of embedding this in the Request input is non-standard to our codebase. You could argue that every field of the `ConnectedObjectsCommand` could/should be part of the `ConnectedObjectsRequest` as well for the same reasons you mention, but yet we don't do that.

A `ConnectedObjectsCommand` represents a specific instance of a struct that is constructed to handle a single request to an endpoint.

If we're going to do this, then we need a larger refactor for consistency. This pattern aligns with the commands/queries structs and the patterns we are using with them as they are designed today.

File: pkg/server/commands/list_objects.go
Comment (jon-whit) [reply pkg/server/commands/list_objects.go:53]: See https://github.com/openfga/openfga/pull/756#discussion_r1199491826

File: internal/graph/check.go
Comment (jon-whit) [reply internal/graph/check.go:383]: It's part of the `typesystem` package though, and that's an internal bit. The server (and by extension commands/queries) layer should be responsible for user facing error messages, since those are the code bits that handle the public API itself.

File: pkg/server/queries.go
Comment (jon-whit) [reply pkg/server/queries.go:100]: Yeah, that's residual. Good catch. That should go away.

File: pkg/typesystem/resolver.go
Comment (jon-whit) [reply pkg/typesystem/resolver.go:33]: It should be. Just missing some code.

File: internal/graph/check.go
Comment (jon-whit) [reply internal/graph/check.go:383]: It's the same reason why the `internal/graph` package doesn't return user facing errors (e.g. errors from the `pkg/server/errors` package).

File: pkg/server/queries.go
Comment (jon-whit) [reply pkg/server/queries.go:1]: Ok, I'd be happy to move it back. But the reason I did it is that it separates the code a bit more, which makes it a little more manageable, and that helped me manage the changes.

Comment (jon-whit) [reply pkg/server/queries.go:1]: Ok, moved back into their original spot for now. We can move those into another file in a subsequent PR.

File: pkg/server/commands/connected_objects.go
Comment (miparnisari) [reply pkg/server/commands/connected_objects.go:105]: > Our pattern is clear - <something>Request structs are for user provided inputs.

This is fine for the requests defined in `api`. But `ResolveCheckRequest`, `ConnectedObjectsRequest` and `reverseExpandRequest` are not defined there, we have more flexibility for these, I think.

> The suggestion of embedding this in the Request input is non-standard to our codebase.

With this specific change you're undoing the change here https://github.com/openfga/openfga/pull/650 and adding more work to yourself. Wouldn't the code work as is today?

Comment (jon-whit) [reply pkg/server/commands/connected_objects.go:105]: > This is fine for the requests defined in api. But ResolveCheckRequest, ConnectedObjectsRequest and reverseExpandRequest are not defined there, we have more flexibility for these, I think.

We are targeting a public API contract for the ConnectedObjects API. Even if we weren't, then it still is a good practice to keep things consistent. Whether an internal RPC or an external RPC it's good to keep consistent practices/semantics between public facing RPC definitions and internal facing ones. 

> With this specific change you're undoing the change here https://github.com/openfga/openfga/pull/650 and adding more work to yourself. Wouldn't the code work as is today?

#650 introduced the inconsistency. All command/query structs are constructed per request, and all other command/query structs define their parameters per struct, not per method. The change, albeit super small and hardly more work for myself, establishes more consistency within the scope of this work.

Comment (rhamzeh) [reply pkg/server/commands/connected_objects.go:105]: (sidenote: if the scope of this discussion is not blocking this PR, let's resolve any other issues, merge and continue the discussion after)

Re: ConnectedObjects
Yes, we do plan to make it public. We need it in order to build a local search index without the user doing all the work.
However:
1) Not sure if `ConnectedObjects` will be the name, that was the name used internally in the API, but it may not be the best name for external usage
2) We need to discuss the interface
As in whatever gets exposed may or may not be this specific method, it will probably be a wrapper on top of it as we debate the API interface. So I wouldn't pre-optimize for that, although it is good to keep in mind that this will end up being exposed in some version or another.
For now, please do not add it to the proto files unless you hide it similar to how we did the UpdateStore, otherwise this will end up being reflected in all SDKs and docs as soon as it is merged.

My thinking is that this should be similar to all the internal methods that CheckRequest and similar wrap, it probably should not have the same inputs (middleware, etc..) that the API entry points do.
Ideally, the external interface would be consistent with the interfaces for the external methods, and the internal one would be consistent with the internal interfaces.

**Comment by jon-whit** [comment]
Closing until we can get around to this at a later time.

**Pull request closed without merging**
